tls/autotls.md

```diff
@@ -0,0 +1,88 @@
+	1. Send a GET request to the `directory` endpoint, and extract the `newAccount` value from the JSON response, which will be the registration URL we'll use
+	2. Signed POST request to registration URL with the following `payload`: `{"termsOfServiceAgreed": true}`. The actual POST body is signed using JWT with an `mykey` and `nonce` (gotten from `directory["newNonce"]`), so the final body of any ACME request should look like:
+		```json
+		{
+		  "payload": token.claims.toBase64,
+		  "protected": token.header.toBase64,
+		  "signature": base64UrlEncode(token.signature),
+		}
+		```
+		Obs: the response to the account registration contains a `kid` in the `location` field that should be saved and used in following requests to ACME server
+5. Request a certificate for the `*.{b36peerid}.libp2p.direct` domain from the ACME server by issuing a POST request using the same JWT signature scheme (and another new `nonce` from `directory["newNonce"]`) but with `kid` instead of `jwk` field and the following payload:
+	```json
+	{
+		"type": "dns",
+		"value": "*.{b36peerid}.libp2p.direct"
+	}
+	```
+6. From the ACME server response, get the entry with `"type"` of `"dns-01"` and derive the `Key Authorization` for it: 
+	-  `sha256.digest((dns01Challenge["token"] + "." + thumbprint(key))` 
+		-  [JWK thumbprint](https://www.rfc-editor.org/rfc/rfc7638): `base64encode(sha256.digest({"e": key.e, "kty": "RSA", "n": key.n}))`, but you can use other key types too
+7. Send challenge to AutoTLS broker/server https://registration.libp2p.direct/, which requires a [PeerID Auth](https://github.com/libp2p/specs/blob/master/http/peer-id-auth.md) scheme:
+	1. Send GET request to the `v1/_acme-challenge` endpoint and get `www-authenticate` field from the response header, and extract the values of three strings that it contains: `challenge-client`, `public-key` and `opaque`
+	2. Generate random string with around 42 characters as a `challengeServer` of our own
+	3. Get `peer-pubkey` and `peer-privkey` keys of our libp2p `peer`, which are not necessarily the same keys we're using to talk to ACME server
+	4. `sig = ` (obs: `varint` is a protobuf [varint](https://protobuf.dev/programming-guides/encoding/#varints) field that encodes the length of the `key=value` string)
+	```
+		sig = base64URL(
+			peer-privkey.sign(
+				bytes(varint + "challenge-client={challenge-client}") +
+				bytes(varint + "hostname={hostname}") +
+				bytes(varint + "server-public-key={publicKey}")
+			)
+		)
+	```
+	5. `headers	 =`
+		```json
+		{
+			"Content-Type": "application/json",
+			"User-Agent": "nim-libp2p",
+			"authorization": "libp2p-PeerID public-key=\"{clientPublicKeyB64}\", opaque=\"{opaque}\", challenge-server=\"{challengeServer}\", sig=\"{sig}\""
+		}
+		```
+	6. Send POST to `v1/_acme-challenge` endpoint using `payload` as body and `headers`
```

Comment:
> using `payload` as body and `headers`.

What is to be done with `headers`? I find the text a bit ambiguous. Maybe something like this instead?
> Send a POST request to the `v1/_acme-challenge` endpoint, using the `payload` as the request body and headers as the HTTP headers.